### PR TITLE
[Issue #704] fix the bug of modelbroadcast when all weights and biases share the one storage

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

fix the bug of modelbroadcast when all weights and biases share the one storage

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/704

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 1]
```diff
@@ -71,6 +71,12 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
         val wb = parameters._1(i)
         weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
           wb.storageOffset(), wb.size(), wb.stride())
+      }
+      i += 1
+    }
+    while (i < parameters._1.length) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: i need init again.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 2]
```diff
@@ -71,6 +71,12 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
...
+      if (parameters._1(i) != null) {
+        val wb = parameters._1(i)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: perhaps parameters._1(i).set() is ok, wb is not necessary

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcastSpec.scala [hunk 3]
```diff
@@ -25,11 +25,27 @@ class ModelBroadcastSpec extends FlatSpec with Matchers {
...
-  val model = LeNet5(10)
+  "model broadcast" should "work properly" in {
+    val sc = new SparkContext(new SparkConf().setMaster("local[1]").setAppName("ModelBroadcast"))
+
+    val model = LeNet5(10)
+
+    val modelBroadCast = ModelBroadcast[Float].broadcast(sc, model)
+    modelBroadCast.value().toString should be(model.toString)
+    modelBroadCast.value().parameters()._1 should be(model.parameters()._1)
+    sc.stop()
```
Comment (yiheng) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcastSpec.scala]: move the sc.stop to
after {
}

to make sure it won't affect other test result when it fail

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcastSpec.scala [hunk 4]
```diff
@@ -25,11 +25,27 @@ class ModelBroadcastSpec extends FlatSpec with Matchers {
...
+  "model broadcast with getParameters" should "work properly" in {
+    val sc = new SparkContext(new SparkConf().setMaster("local[1]").setAppName("ModelBroadcast"))
+
+    val model = LeNet5(10)
+    model.getParameters()
+
+    val modelBroadCast = ModelBroadcast[Float].broadcast(sc, model)
+    modelBroadCast.value().toString should be(model.toString)
+    modelBroadCast.value().parameters()._1 should be(model.parameters()._1)
+    sc.stop()
```
Comment (yiheng) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcastSpec.scala]: move the sc.stop to
after {
}

to make sure it won't affect other test result when it fail

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 5]
```diff
@@ -71,6 +71,12 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
         val wb = parameters._1(i)
         weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
           wb.storageOffset(), wb.size(), wb.stride())
+      }
+      i += 1
+    }
+    while (i < parameters._1.length) {
+      if (parameters._1(i) != null) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: parameters._2(i) != null

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 6]
```diff
@@ -71,7 +71,13 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
         val wb = parameters._1(i)
         weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
           wb.storageOffset(), wb.size(), wb.stride())
-        wb.set()
+      }
+      i += 1
+    }
+    i = 0
+    while (i < parameters._1.length) {
+      if (parameters._2(i) != null) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: should be parameters._2(i) != null && parameters._1(i) != null, right?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 7]
```diff
@@ -71,7 +71,13 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
...
+        parameters._1(i).set()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: I don't think use the parameters() is a good choice, why not use getParameters?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 8]
```diff
@@ -71,7 +71,13 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
         val wb = parameters._1(i)
         weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
           wb.storageOffset(), wb.size(), wb.stride())
-        wb.set()
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: I don't understand - why resetting a tensor will impact another tensor? If so, this seems to be a bug of tensor

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: reset will reset the storage, while all the weight/bias tensor sharing the same storage(with different offset)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala (hunk 7 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: getParameters will copy all the parameters, which may not be necessary if only the inference is needed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 9]
```diff
@@ -71,7 +71,15 @@ class ModelBroadcast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Seriali
         val wb = parameters._1(i)
         weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: Why we need this? can we change to wb.clone()?

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: The weights are shared among models when prediction.

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: @SeaOfOcean I think when you add the optimization, you should add comments to explain it. Or people will get confused. Can you add some comments?

**Pull request merged**